### DISCUSSION TITLE
Redirect not working when using form actions inside onSubmit handlers

### DISCUSSION THREAD

**Discussion by anthonyalayo** [vercel/next.js#73199] [Help]
### Summary

I have a multi-step form across multiple components. I am using controlled components in order to keep it all in memory, and I have to submit via `onSubmit` instead of `action`, since not all the input fields are on the screen at the time of submission.

I have attempted to do something like this:

    <form
      onSubmit={form.onSubmit(async (values, event) => {
        const formData = new FormData();

        for (const [key, value] of Object.entries(values)) {
          formData.append(key, value);
        }

        startTransition(async () => {
          await formAction(formData);
        });
      })}
    >

However, I get this weird error on my IDE when calling the formAction:
```
TS80007: await has no effect on the type of this expression.
```

Most likely related to the above error, my redirect in my server action does not work. It only works when it is being called inside `action`. How does one make this work?

**Comment by icyJoseph** [comment]
Assuming, `formAction` comes from `useActionState`, then it makes sense that you cannot await it. `useActionState` handles the "wiring" for you, so it does the transition, and awaits the result of your server action, etc... you just have to call the trigger function, which is itself not a promise.

Also, a side note about Object.entries and FormData, https://x.com/rwieruch/status/1857105224806068629

As to why it doesn't redirect... I am not sure, I remember seeing issues that redirect was not working correctly, but that Next 15 fixed it?

**Reply by anthonyalayo** [reply]
@icyJoseph thanks for the response, is what I'm doing correct? How does one handle the server action response?

**Reply by icyJoseph** [reply]
Right, like, if a fetch request, has a response with redirect, the API automatically, follows that, without navigating the document. In other words, your code gets the response obtained by following the redirect.

A form submission is seen as, the document submitting the request, so the document follows the redirect. 

Are you using useActionState? What state do you have after attempting to redirect?

**Reply by icyJoseph** [reply]
See here, https://github.com/vercel/next.js/blob/c50109b4468bb274c29dfc0eb5aaa906736dc4d9/packages%2Fnext%2Fsrc%2Fclient%2Fcomponents%2Frouter-reducer%2Freducers%2Fserver-action-reducer.ts#L88-L110 

They do some trickery to trigger an error boundary, and redirect the document from there. 

Still strange... Any more info, logs, headers we could look at?

**Reply by anthonyalayo** [reply]
I hacked this together, what do you think of this? I'm quietly injecting all of the fields that are missing on the last step of my multi-step form before I trigger the action. That way, everything works as-is without other complications of handling errors / function params / etc differently.

```
export function injectFieldsAndRequestSubmit(
  values: Record<string, any>,
  event: React.FormEvent<HTMLFormElement> | undefined
) {
  const formRef = event?.currentTarget;
  if (!formRef) {
    return;
  }
  for (const [key, value] of Object.entries(values)) {
    let input = formRef.querySelector(`[name="${key}"]`) as HTMLInputElement | null;

    if (!input) {
      input = document.createElement('input');
      input.type = 'hidden';
      input.name = key;
      formRef.appendChild(input);
    }

    if (value) {
      input.value = value.toString();
    }
  }
  event?.currentTarget.requestSubmit();
}
..
..
  const [, formAction, pending] = useActionState(signUp, { error: '' });
...
...
        <form onSubmit={form.onSubmit(injectFieldsAndRequestSubmit)} action={formAction}>
```
